### PR TITLE
Convert Request, Response, CodeResponseType, TokenResponseType to ES6 classes

### PR DESCRIPTION
**Pull request by menewman**
## Summary

Chipping away at #212 by converting four more classes:
- Request
- Response
- CodeResponseType
- TokenResponseType

(Do note that TokenResponseType's constructor simply throws a not implemented error, though. So that conversion was easy. 😉)

## Linked issue(s)
A few checkboxes of #212 (but not the whole thing).

## Involved parts of the project

Specifically the classes noted above, which affect pretty much everything (since Request/Response are core parts of the interface). But, there are no functional changes.

## Added tests?

I did not update the tests, but these files are already covered under the existing tests, which should continue to pass after this refactoring.

## OAuth2 standard

N/A

## Reproduction

N/A

### CODE REVIEW

**Review by jorenvandeweyer — CHANGES_REQUESTED** [review]
@menewman I got some small comments, could you please update these?

@jankapunkt PR seems fine. Could you also review this so this one can be merged?

File: lib/request.js
Comment (jankapunkt) [reply lib/request.js:36]: We need to use `Object.prototype.hasOwnProperty.call(options, property)` anyway here, to avoid manipulation of the prototype chain, especially because Request options may be built from a client.

File: lib/request.js
Comment (jankapunkt) [reply lib/request.js:56]: @jorenvandeweyer is there ever a case where types would be a nested array?

File: lib/response.js
Comment (jankapunkt) [inline lib/response.js:16]: I propose to use `Object.prototype.hasOwnProperty.call` here as well to avoid any tampering with the prototype chain

**Review by jankapunkt — CHANGES_REQUESTED** [review]
There are some minor changes but I think at keast to has own property checks should be considered, even if the attack surface is rather superficial and imaginary

File: lib/request.js
Comment (jorenvandeweyer) [reply lib/request.js:56]: The passed types would not be a nested array. By using the spread operator an array would be converted to a nested array.

```js
request.is(['a', 'b']) // types = [['a', 'b']]
request.is('a', 'b') // types = ['a', 'b']
```

Comment (jankapunkt) [reply lib/request.js:56]: I see, makes sense

File: lib/request.js
Comment (menewman) [reply lib/request.js:36]: @jankapunkt - `Object.entries()` (like `Object.keys()`) does not traverse the prototype chain but only iterates over the own enumerable properties of the object.

[MDN docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/entries#description):
> `Object.entries()` returns an array whose elements are arrays corresponding to the enumerable string-keyed property key-value pairs found directly upon object. This is the same as iterating with a [for...in](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in) loop, except that a for...in loop enumerates properties in the prototype chain as well. The order of the array returned by Object.entries() is the same as that provided by a [for...in](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in) loop.

In other words, I believe the `hasOwnProperty` check will always return true for anything we iterate other with `Object.entries()`, and therefore it is redundant/not necessary. It is certainly possible I'm misunderstanding the concern you have in mind, though, so please let me know if I'm missing the mark!

Comment (menewman) [reply lib/request.js:36]: @jorenvandeweyer - You're right that it's not strictly necessary to check again; I had the same thought while I was refactoring.

The reason I ultimately left the existing check in place is because I worried that trusting the destructuring as a safeguard is a bit brittle -- I could easily see someone in the future changing which options are destructured without realizing that it could have a side effect on the "custom property" logic.

I'm happy to defer to the maintainers' preference, though. So, let me know if you would prefer for me to remove the check, and I'll do it.

File: lib/request.js
Comment (menewman) [reply lib/request.js]: Done! Happy to keep it more concise.

**Review by jorenvandeweyer — APPROVED** [review]

File: lib/request.js
Comment (jankapunkt) [reply lib/request.js:36]: @menewman you are correct but take a close look which own properties are used by Object.entries - the ones from the options but the assignment goes to the request instance.

Therefore I could tamper properties from the chain like this:

```js
const req = new Request({ body, headers, toString: () => '[object IAMValid]' })
req.toString() // [object IAMValid]
```

This is because `toString` is an own property of the options but for request it's part of the prototype chain. This could theoretically allow to tamper the chain.

I know this is very superficial but I think we are on the safe side, adding the extra check

Comment (jankapunkt) [reply lib/request.js:36]: @jorenvandeweyer I will add this to the next PR for the improved tests (#219)

Comment (menewman) [reply lib/request.js:36]: @jankapunkt - I don't think that's possible because of the `.filter(([property]) => !this[property])` check.

With the combo of Object.entries + this filter function...
- We only loop over enumerable keys on options
- We only overwrite keys on `this` that do not already exist (including directly on the object, or on the prototype chain)

I believe this is functionally identical to the prior version of the logic with the for...in + hasOwnProperty check.

But the best proof is always in tests, so I also added a couple of unit tests to check this case specifically in my latest commit!

Comment (menewman) [reply lib/request.js:36]: https://github.com/node-oauth/node-oauth2-server/pull/225/commits/8ea66994831892dc28ec0215246ecd2a3752f08c

Comment (jankapunkt) [reply lib/request.js:36]: @menewman thank you I understand it now and thanks for the added tests

**Review by jankapunkt — APPROVED** [review]

**Pull request merged**
